### PR TITLE
test: add workspace tests

### PR DESCRIPTION
**Pull request by johnkim-det**
## Ticket

CM-466

## Description

Adds e2e tests to cover filtering, sorting, and toggling the view on the Workspaces page. Also fixes some existing edge case issues (using correct user to create workspaces, testId for empty message components).

## Test Plan

No additional release party testing required. Automated tests should pass.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/components/GridListRadioGroup.tsx [hunk 1]
```diff
@@ -24,15 +24,17 @@ const GridListRadioGroup: React.FC<Props> = ({ onChange, value }: Props) => {
...
-    <RadioGroup
-      iconOnly
-      options={[
-        { icon: 'grid', id: GridListView.Grid, label: 'Grid View' },
-        { icon: 'list', id: GridListView.List, label: 'List View' },
-      ]}
-      value={value}
-      onChange={handleChange}
-    />
+    <div data-testid="grid-list-radio-group">
```
Comment (JComins000) [inline webui/react/src/components/GridListRadioGroup.tsx:27]: in general, i prefer data-test-component as a top level testid in case another `data-testid` get added when `RadioGroup` is implemented. nbd either way

File: webui/react/src/e2e/fixtures/global-fixtures.ts [hunk 2]
```diff
@@ -192,11 +192,7 @@ export const test = baseTest.extend<CustomFixtures, CustomWorkerFixtures>({
         },
       });
       const adminUser = await backgroundApiUser.createUser(request);
-      await backgroundApiUser.apiAuth.loginApi({
```
Comment (JComins000) [inline webui/react/src/e2e/fixtures/global-fixtures.ts:195]: 👍

File: webui/react/src/e2e/models/pages/WorkspaceList.ts [hunk 3]
```diff
@@ -36,7 +52,18 @@ export class WorkspaceList extends DeterminedPage {
     parent: this,
     selector: '[data-testid="newWorkspace"]',
   });
-  // TODO missing grid list
+  readonly gridListRadioGroup = new GridListRadioGroup({
+    parent: this,
+    selector: '[data-testid="grid-list-radio-group"]',
```
Comment (JComins000) [inline webui/react/src/e2e/models/pages/WorkspaceList.ts:57]: remove this line

File: webui/react/src/e2e/models/components/GridListRadioGroup.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { BaseComponent } from 'playwright-page-model-base/BaseComponent';
+
+export class GridListRadioGroup extends BaseComponent {
```
Comment (JComins000) [inline webui/react/src/e2e/models/components/GridListRadioGroup.ts:3]: change this to named component with defaultSelector = `[data-testid="grid-list-radio-group"]`

File: webui/react/src/e2e/models/pages/WorkspaceList.ts [hunk 5]
```diff
@@ -36,7 +52,18 @@ export class WorkspaceList extends DeterminedPage {
...
+  });
+  readonly table = new InteractiveTable({
+    parent: this,
+    tableArgs: {
+      attachment: '[data-testid="table"]',
```
Comment (JComins000) [inline webui/react/src/e2e/models/pages/WorkspaceList.ts:62]: just fyi, this has `  readonly defaultSelector = 'div[data-test-component="interactiveTable"]';`. if i had used data-testid instead, then it would've conflicted here. that's why i like to use data-test-component, in case a conflict might arise.

File: webui/react/src/e2e/tests/projectsWorkspaces.spec.ts [hunk 6]
```diff
@@ -115,10 +129,10 @@ test.describe('Workspace UI CRUD', () => {
     });
   });
 
-  test('Pin and Unpin a Workspace from Card', async ({ authedPage, backgroundApiWorkspace }) => {
+  test('Pin and Unpin a Workspace from Card', async ({ authedPage, apiWorkspace }) => {
```
Comment (JComins000) [inline webui/react/src/e2e/tests/projectsWorkspaces.spec.ts:132]: 👍

File: webui/react/src/e2e/tests/projectsWorkspaces.spec.ts [hunk 7]
```diff
@@ -239,32 +247,143 @@ test.describe('Workspace UI CRUD', () => {
...
-
-    await test.step('Navigate to Workspaces', async () => {
-      await workspaceList.goto();
-      await workspaceCard.pwLocator.waitFor({ timeout: 10_000 });
-    });
-  });
 
   test('Attempt to delete a workspace but with bad validation', async ({
     authedPage,
-    newWorkspace,
```
Comment (JComins000) [inline webui/react/src/e2e/tests/projectsWorkspaces.spec.ts:257]: this thing took care of teardown. does this workspace ever get torn down?

File: webui/react/src/e2e/tests/projectsWorkspaces.spec.ts [hunk 8]
```diff
@@ -239,32 +247,143 @@ test.describe('Workspace UI CRUD', () => {
...
+    await workspaceList.gridListRadioGroup.list.pwLocator.click();
+
+    const idSortedWorkspaceNames = _.orderBy(workspaces, 'id', 'desc').map((w) => w.name);
+
+    expect(await workspaceList.table.table.rows.nth(0).name.pwLocator.textContent()).toEqual(
+      idSortedWorkspaceNames[0],
+    );
+
+    // cleanup:
+    await workspaceList.gridListRadioGroup.grid.pwLocator.click();
```
Comment (JComins000) [inline webui/react/src/e2e/tests/projectsWorkspaces.spec.ts:373]: can you move this to beforeEach?

Comment (JComins000) [reply webui/react/src/e2e/tests/projectsWorkspaces.spec.ts:373]: might be best to make a module level beforeEach for this and that sort beforeEach thing we added

File: webui/react/src/pages/WorkspaceList.tsx [hunk 9]
```diff
@@ -394,15 +394,15 @@ const WorkspaceList: React.FC = () => {
           workspacesList
         ) : settings.whose === WhoseWorkspaces.All && settings.archived && !isLoading ? (
           <Message
-            data-testid="noWorkspaces"
             description={`Create a workspace to keep track of related projects and ${f_flat_runs ? 'searches' : 'experiments'}.`}
             icon="workspaces"
+            testId="noWorkspaces"
```
Comment (JComins000) [inline webui/react/src/pages/WorkspaceList.tsx:399]: did the models get updated too?

File: webui/react/src/pages/WorkspaceList.tsx [hunk 10]
```diff
@@ -394,15 +394,15 @@ const WorkspaceList: React.FC = () => {
...
             title="No Workspaces"
           />
         ) : (
           <Message
-            data-testid="noMatchingWorkspaces"
             icon="warning"
+            testId="noMatchingWorkspaces"
```
Comment (JComins000) [inline webui/react/src/pages/WorkspaceList.tsx:405]: did the models get updated too?

File: webui/react/src/pages/WorkspaceList.tsx [hunk 11]
```diff
@@ -178,7 +178,7 @@ const WorkspaceList: React.FC = () => {
         dataIndex: 'name',
         defaultWidth: DEFAULT_COLUMN_WIDTHS['name'],
         key: V1GetWorkspacesRequestSortBy.NAME,
-        onCell: onRightClickableCell,
+        onCell: () => ({ ...onRightClickableCell(), 'data-testid': 'name' }),
```
Comment (JComins000) [inline webui/react/src/pages/WorkspaceList.tsx:181]: actually, i like `data-test-value` on variable test hooks. nbd either way

File: webui/react/src/e2e/tests/projectsWorkspaces.spec.ts (hunk 8 above)
Comment (johnkim-det) [reply webui/react/src/e2e/tests/projectsWorkspaces.spec.ts]: actually I realized that this is only needed to support the `beforeEach` in `With a Workspace`, which I actually don't think is necessary.

File: webui/react/src/e2e/tests/projectsWorkspaces.spec.ts (hunk 7 above)
Comment (johnkim-det) [reply webui/react/src/e2e/tests/projectsWorkspaces.spec.ts:257]: the id gets pushed to the `workspaceIds` which are deleted in the `afterAll`

File: webui/react/src/pages/WorkspaceList.tsx (hunk 10 above)
Comment (johnkim-det) [reply webui/react/src/pages/WorkspaceList.tsx:405]: don't think they needed to be updated, they were already looking for this testId, but the hew component was not rendering with it.

File: webui/react/src/pages/WorkspaceList.tsx (hunk 11 above)
Comment (johnkim-det) [reply webui/react/src/pages/WorkspaceList.tsx:181]: think I'll keep this as is for now just because it's already the convention in `UserManagement.tsx`

**Review by JComins000 — APPROVED** [review]

File: webui/react/src/e2e/models/pages/WorkspaceList.ts [hunk 12]
```diff
@@ -4,10 +4,26 @@ import { DeterminedPage } from 'e2e/models/common/base/BasePage';
...
+  readonly name = new BaseComponent({
+    parent: this,
+    selector: '[data-testid="name"]',
+  });
+}
+
+class WorkspaceRow extends Row {
+  readonly name = new BaseComponent({
+    parent: this,
+    selector: '[data-testid="name"]',
```
[the remark this answers is not in the document]
Comment (JComins000) [reply webui/react/src/e2e/models/pages/WorkspaceList.ts:23]: I like to use the single approach because it makes for good references. The approach I chose comes from my experience, but if perhaps Hew is expecting something else, then it's okay to have exceptions.
